bigquery_etl/routine/parse_routine.py

```diff
@@ -144,186 +158,162 @@ def from_text(text, dataset, name, filepath=None, description="", is_defined=Tru
+        return RawRoutine(
             internal_name,
             dataset,
             filepath,
             definitions,
             tests,
             # We convert the list to a set to deduplicate entries,
             # but then convert back to a list for stable order.
-            sorted(set(dependencies)),
+            sorted(dependencies),
             description,
             is_stored_procedure,
         )
 
 
 @dataclass
-class ParsedUdf(RawUdf):
-    """Parsed representation of a UDF including dependent UDF code."""
+class ParsedRoutine(RawRoutine):
+    """Parsed representation of a routine including dependent routine code."""
 
     tests_full_sql: List[str]
 
     @staticmethod
-    def from_raw(raw_udf, tests_full_sql):
-        """Promote a RawUdf to a ParsedUdf."""
-        return ParsedUdf(*astuple(raw_udf), tests_full_sql)
-
-
-def get_udf_dirs(udf_dirs, project_id=None):
-    """Return paths to directories with UDFs."""
-    # todo: make this more generic, I don't think we need to make the distinction
-    # between mozfun and other project here since UDFs now live in project directories
-    if project_id != "mozfun":
-        # for non-mozfun projects, the default UDF directories are udf/ and udf_js/
-        # the project needs to be pre-pended to these paths
-        projects = project_dirs(project_id)
-
-        udf_dirs = [
-            os.path.join(project, d)
-            for project in projects
-            for d in udf_dirs
-            if os.path.exists(os.path.join(project, d))
-        ]
+    def from_raw(raw_routine, tests_full_sql):
+        """Promote a RawRoutine to a ParsedRoutine."""
+        return ParsedRoutine(*astuple(raw_routine), tests_full_sql)
 
-    return udf_dirs
 
-
-def read_udf_dirs(*udf_dirs):
-    """Read contents of udf_dirs into dict of RawUdf instances."""
+def read_routine_dir(project_dir):
+    """Read contents of routine dirs into dict of RawRoutine instances."""
     return {
-        raw_udf.name: raw_udf
-        for udf_dir in (udf_dirs or get_udf_dirs(UDF_DIRS) + list(MOZFUN_DIR))
-        for root, dirs, files in os.walk(udf_dir)
+        raw_routine.name: raw_routine
+        for root, dirs, files in os.walk(project_dir)
         if os.path.basename(root) != EXAMPLE_DIR
         for filename in files
         if filename in (UDF_FILE, PROCEDURE_FILE)
-        for raw_udf in (RawUdf.from_file(os.path.join(root, filename)),)
+        for raw_routine in (RawRoutine.from_file(os.path.join(root, filename)),)
     }
 
 
-def parse_udf_dirs(*udf_dirs):
-    """Read contents of udf_dirs into ParsedUdf instances."""
+def parse_routines(project_dir):
+    """Read routine contents of the project dir into ParsedRoutine instances."""
     # collect udfs to parse
-    raw_udfs = read_udf_dirs(*udf_dirs)
+    raw_routines = read_routine_dir(project_dir)
+    raw_routines.update(read_routine_dir(os.path.join(SQL_DIR, "mozfun")))
+    raw_routines.update(read_routine_dir(ASSERT_UDF_DIR))
```

Comment:
These are all UDFs that are potentially referenced in tests

Reply:
It's looking like we should allow a list of routine dirs: `read_routine_dirs(*project_dirs)`